src/engine/input/key-mapping/InputGroupList.cs

```diff
@@ -0,0 +1,180 @@
+﻿using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using Godot;
+using Newtonsoft.Json;
+
+/// <summary>
+///   Used by OptionsMenu>Inputs>InputGroupContainer
+/// </summary>
+public class InputGroupList : VBoxContainer
+{
+    [Export]
+    public NodePath ConflictDialogPath;
+
+    internal static readonly PackedScene InputEventItemScene =
+        GD.Load<PackedScene>("res://src/engine/input/key-mapping/InputEventItem.tscn");
+
+    private static readonly Dictionary<string, List<InputEventWithModifiers>> DefaultControlScheme =
+        GetCurrentlyAppliedControlScheme();
+
+    private ConfirmationDialog conflictDialog;
+
+    private InputEventItem latestDialogCaller;
+    private InputEventItem latestDialogConflict;
+    private InputEventWithModifiers latestDialogNewEvent;
+
+    public InputGroupList()
+    {
+        Instance = this;
+    }
+
+    public delegate void ControlSchemeChangedDelegate(Dictionary<string, List<InputEventWithModifiers>> data);
+
+    public event ControlSchemeChangedDelegate OnControlSchemeChanged;
+
+    public static InputGroupList Instance { get; private set; }
+    public IEnumerable<InputGroupItem> AllGroupItems => GetChildren().OfType<InputGroupItem>();
+
+    public static Dictionary<string, List<InputEventWithModifiers>> GetDefaultControlScheme()
+    {
+        return DefaultControlScheme;
+    }
+
+    public static Dictionary<string, List<InputEventWithModifiers>> GetCurrentlyAppliedControlScheme()
+    {
+        return InputMap.GetActions().OfType<string>()
+            .ToDictionary(p => p,
+                p => InputMap.GetActionList(p).OfType<InputEventWithModifiers>()
+                    .ToList());
+    }
```

Comment:
This is really complex? Why is it so complex to know the current controls?

Reply:
I have not found a better solution to get the current godot inputs

Reply:
Then could you write a wrapper class or something?
Reflecting back, I think that maybe you should rework your code to be way less tightly coupled to Godot's InputMap by introducing wrapper classes that are used in place of the InputMap, until it is time to apply the final results.